CMake fails when CMAKE_BINARY_DIR contains '+'.

Getting the following error message:

```
RegularExpression::compile(): Nested *?+.
RegularExpression::compile(): Error in compile.
RegularExpression::compile(): Nested *?+.
RegularExpression::compile(): Error in compile.
CMake Error at cmake/HPX_GeneratePackage.cmake:56 (if):
  if given arguments:

    "(" "NOT" "dir" "MATCHES" "^/home/ibbles/mirrored/programmering/others/c++/hpx/hpx_0.9.10_build.*" ")" "AND" "(" "NOT" "dir" "MATCHES" "^/home/ibbles/mirrored/programmering/others/c++/hpx/hpx_0.9.10_repo.*" ")"

  Regular expression
  "^/home/ibbles/mirrored/programmering/others/c++/hpx/hpx_0.9.10_repo.*"
  cannot compile
Call Stack (most recent call first):
  CMakeLists.txt:1350 (include)


-- Configuring incomplete, errors occurred!
```

Renaming the folder 'c++' to 'cpp' will let CMake run successfully.

--- a/cmake/HPX_GeneratePackage.cmake
+++ b/cmake/HPX_GeneratePackage.cmake
@@ -52,9 +52,21 @@ endforeach()
 
 # Get the include directories we need ...
 get_directory_property(_INCLUDE_DIRS INCLUDE_DIRECTORIES)
+
+# replace all characters with special regex meaning
+set(special_chars "^;+;*;?;$;.;-;|;(;);]")
+set(binarydir_escaped ${CMAKE_BINARY_DIR})
+set(sourcedir_escaped ${PROJECT_SOURCE_DIR})
+foreach(special_char ${special_chars})
+  string(REPLACE "${special_char}" "\\${special_char}" binarydir_escaped ${binarydir_escaped})
+  string(REPLACE "${special_char}" "\\${special_char}" sourcedir_escaped ${sourcedir_escaped})
+endforeach()
+
+# '[' has special meaning in lists
+string(REPLACE "[" "\\[" binarydir_escaped ${binarydir_escaped})
+string(REPLACE "[" "\\[" sourcedir_escaped ${sourcedir_escaped})
+
 foreach(dir ${_INCLUDE_DIRS})
-  string(REPLACE "++" "\\+\\+" binarydir_escaped ${CMAKE_BINARY_DIR})
-  string(REPLACE "++" "\\+\\+" sourcedir_escaped ${PROJECT_SOURCE_DIR})
   if((NOT dir MATCHES "^${binarydir_escaped}.*")
     AND (NOT dir MATCHES "^${sourcedir_escaped}.*"))
     set(_NEEDED_INCLUDE_DIRS "${_NEEDED_INCLUDE_DIRS} -I${dir}")

